x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RepositorySpeedTestAction.java

```diff
@@ -0,0 +1,718 @@
+                            return new BlobSpeedTestAction.Response(in);
+                        }
+                    }
+                );
+            }
+
+        }
+
+        private BlobContainer getBlobContainer() {
+            return repository.blobStore().blobContainer(new BlobPath().add(blobPath));
+        }
+
+        private void onWorkerCompletion() {
+            tryListAndCleanup(0);
+        }
+
+        private void tryListAndCleanup(final int retryCount) {
+            if (timeoutTimeMillis < currentTimeMillisSupplier.getAsLong() || task.isCancelled()) {
+                logger.warn(
+                    "speed test of repository [{}] failed in cleanup phase after [{}] attempts, attempting best-effort cleanup "
+                        + "but you may need to manually remove [{}]",
+                    request.getRepositoryName(),
+                    retryCount,
+                    blobPath
+                );
+                isRunning(); // set failure if not already set
+                deleteContainerAndSendResponse();
+            } else {
+                boolean retry = true;
+                try {
+                    if (listingStartTimeNanos.isEmpty()) {
+                        assert retryCount == 0 : retryCount;
+                        logger.trace(
+                            "all tasks completed, checking expected blobs exist in [{}:{}] before cleanup",
+                            request.repositoryName,
+                            blobPath
+                        );
+                        listingStartTimeNanos = OptionalLong.of(System.nanoTime());
+                    } else {
+                        logger.trace(
+                            "retrying check that expected blobs exist in [{}:{}] before cleanup, retry count = {}",
+                            request.repositoryName,
+                            blobPath,
+                            retryCount
+                        );
+                    }
+                    final BlobContainer blobContainer = getBlobContainer();
+                    final Map<String, BlobMetadata> blobsMap = blobContainer.listBlobs();
+
+                    final Set<String> missingBlobs = new HashSet<>(expectedBlobs);
+                    missingBlobs.removeAll(blobsMap.keySet());
+                    if (missingBlobs.isEmpty()) {
+                        logger.trace(
+                            "all expected blobs found after {} failed attempts, cleaning up [{}:{}]",
+                            retryCount,
+                            request.getRepositoryName(),
+                            blobPath
+                        );
+                        retry = false;
+                        deleteContainerAndSendResponse();
+                    } else {
+                        logger.debug(
+                            "expected blobs [{}] missing in [{}:{}], trying again; retry count = {}",
```

Comment:
Do we need this any more? AFAIK S3 was the only blob store that didn't have consistent listings, but it does now doesn't it?

Reply:
I think we should require consistent listings (as well as test for it), as all currently officially supported blob stores support it.